match/tensordata.py

```diff
@@ -0,0 +1,154 @@
+import itertools
+import math
+from typing import Union
+
+
+class TensorData:
+    def __init__(self, *size: int, value: float = 0.0, dtype: type = float):
+        self.shape: tuple[int] = size
+        self.dtype: type = dtype
+        self.__initialize_tensor_data(value)
+        self.__initialize_strides()
+
+    def __initialize_strides(self):
+        if not self._data:
+            self._strides = ()
+            return
+
+        strides = []
+        current_stride = len(self._data)
+        for dim in self.shape:
+            current_stride /= dim
+            strides.append(current_stride)
+        self._strides = tuple(strides)
+
+    def __initialize_tensor_data(self, value):
+        num_elements = 1 if self.shape else 0
+        for dim in self.shape:
+            num_elements *= dim
+        self._item = None if self.shape else value
+        self._data = [TensorData(value=value) for _ in range(num_elements)]
+
+    def __out_of_bounds_coords(self, coords: tuple):
+        if len(coords) != len(self.shape):
+            raise IndexError(f"Too many dimensions, expected {len(self.shape)}.")
+        for i, j in zip(coords, self.shape):
+            if i < 0 or i >= j:
```

Comment:
Is `i == j` valid?

Reply:
No. In my code, a dimension of 3 implies that only indices 0,1,2 are in bounds. Is this what you mean?

Reply:
My mistake. I read it as checking for correct values instead of incorrect. Please ignore this and the one below.